grails-app/controllers/org/pih/warehouse/inventory/StockMovementController.groovy

```diff
@@ -283,6 +287,95 @@ class StockMovementController {
+            redirect(action: "show", id: stockMovement?.id)
+            return
+        }
+        if (comment.sender.id != session.user.id) {
+            flash.message = "${g.message(code: 'auth.notAuthorized.message')}"
+            redirect(action: "show", id: stockMovement?.id)
+            return
+        }
+        render(view: "addComment", model: [stockMovement: stockMovement, comment: comment])
+    }
+
+    def deleteComment = {
+        def stockMovement = getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'Stock Movement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (!comment) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'comment.label', default: 'Comment'), params.id])}"
+            redirect(action: "show", id: stockMovement?.id)
+            return
+        }
+
+        if (comment.sender.id != session.user.id) {
+            throw new AccessDeniedException("You are not authorized to perform this action")
+        }
+
+        stockMovement.requisition?.removeFromComments(comment)
+        if (!stockMovement.requisition.hasErrors() && stockMovement.requisition?.save()) {
+            flash.message = "${g.message(code: 'default.comments.deleted.label')}"
+            redirect(action: "show", id: stockMovement.id)
+            return
+        }
+        render(view: "show", model: [stockMovement: stockMovement])
+    }
+
+    def saveComment = {
+        def stockMovement = getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (comment) {
+            if (comment.sender.id != session.user.id) {
+                throw new AccessDeniedException("You are not authorized to perform this action")
+            }
+            comment.properties = params
+            if (!stockMovement.requisition.hasErrors() && stockMovement.requisition.save()) {
+                flash.message = "${g.message(code: 'default.comments.updated.label')}"
+                redirect(action: "show", id: stockMovement.id)
+                return
+            }
+        } else {
+            comment = new Comment(params)
+            stockMovement.requisition.addToComments(comment)
+            if (!stockMovement.requisition.hasErrors() && stockMovement.requisition.save()) {
```

Comment:
Should it have a `flush`? I think this controller is not transacitonal

Reply:
in my opinion most of this method should be a service method, especially the save calls.
Having this code, it would:

- break in Grails 3 when using flush immedietaly in this line (no transaction in progress error)
- break "silently" (or at the end of the session) in Grails 3 when not using flush (the error would happen "in the background")
- it will work fine in Grails 1 due to the OSIV

so it should be refactored in my opinion and at least moved to the service.

Moreover I think we used to use `update` method when we update an entity, and when we create new, we used to use `create` method, don't we?

Reply:
Agree wholeheartedly with the "move to service layer" suggestion.